src/Raven.Embedded/EmbeddedServer.cs

```diff
@@ -275,37 +275,34 @@ private void ShutdownServerProcess(Process process)
             string? url = null;
             var startupDuration = Stopwatch.StartNew();
 
-            var outputString = await ProcessHelper.ReadOutput(process.StandardOutput, startupDuration, _serverOptions, async (line, builder) =>
+            var stderrBuilder = new StringBuilder();
+            process.ErrorDataReceived += (_, receivedEventArgs) =>
             {
-                if (line == null)
-                {
-                    var errorString = await ProcessHelper.ReadOutput(process.StandardError, startupDuration, _serverOptions, null).ConfigureAwait(false);
-
-                    ShutdownServerProcess(process);
-
-                    throw new InvalidOperationException(BuildStartupExceptionMessage(builder.ToString(), errorString));
-                }
+                if (receivedEventArgs.Data != null)
+                    stderrBuilder.AppendLine(receivedEventArgs.Data);
+            };
+            process.BeginErrorReadLine();
 
+            var stdoutString = await ProcessHelper.ReadOutput(process.StandardOutput, startupDuration, _serverOptions, (line, _) =>
+            {
                 const string prefix = "Server available on: ";
-                if (line.StartsWith(prefix))
-                {
-                    url = line.Substring(prefix.Length);
-                    return true;
-                }
+                if (line.StartsWith(prefix) == false)
+                    return Task.FromResult(false);
+
+                url = line.Substring(prefix.Length);
+                return Task.FromResult(true);
 
-                return false;
             }).ConfigureAwait(false);
 
-            if (url == null)
-            {
-                var errorString = await ProcessHelper.ReadOutput(process.StandardError, startupDuration, _serverOptions, null).ConfigureAwait(false);
+            if (url != null)
+                return (ServerUrl: new Uri(url), process);
 
-                ShutdownServerProcess(process);
+            process.CancelErrorRead();
+            var stderrString = stderrBuilder.ToString();
```

Comment:
Do we need to worry about concurrency here?
So we are _adding_ an error line - and you are calling `ToString()` at the same time. Pretty sure that `CancelErrorLine` isn't going to wait for all pending stuff.

Reply:
Thank you for bringing this to my attention.
I refactored this section of code and now I'm expecting either the stream to close because `receivedEventArgs.Data == null` (as an "EOF-like" marker) or I'm waiting 5 seconds and outputting everything the string builder had accumulated up to that point.